x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java

```diff
@@ -408,18 +407,6 @@ public void testSameExpressionWithoutAlias() {
         assertNotEquals(eqe.output().get(0).id(), eqe.output().get(1).id());
     }
 
-    public void testInOutOfRangeValues() {
```

Comment:
These two tests assert a behavior that's different between `IN` and `=`: `IN` tries to convert each value in the list to the type on the left hand side. `=` on the other hand does not perform this conversion and passes non-matching types to ES. So both `select emp_no from test_emp where emp_no in (9223372036854775807)` and `select emp_no from test_emp where emp_no < 9223372036854775807` fail, but the former with
```
{"error":{"root_cause":[{"type":"ql_illegal_argument_exception","reason":"[9223372036854775807] out of [integer] range"}],"type":"ql_illegal_argument_exception","reason":"[9223372036854775807] out of [integer] range"},"status":500}
```
and the latter with
```
{"error":{"root_cause":[{"type":"query_shard_exception","reason":"failed to create query: Value [9223372036854775807] is out of range for an integer","index_uuid":"JrctQS1KTXmIqlD6_vkAQQ","index":"test_emp"}],"type":"search_phase_execution_exception","reason":"all shards failed","phase":"query","grouped":true,"failed_shards":[{"shard":0,"index":"test_emp","node":"6a6CBLYYQFa6gqsxbMkn5Q","reason":{"type":"query_shard_exception","reason":"failed to create query: Value [9223372036854775807] is out of range for an integer","index_uuid":"JrctQS1KTXmIqlD6_vkAQQ","index":"test_emp","caused_by":{"type":"illegal_argument_exception","reason":"Value [9223372036854775807] is out of range for an integer"}}}]},"status":400}
```

I now aligned the behavior such that both `IN` and `=` pass the values to ES.

Another alternative would have been to adopt the `IN` behavior and do the check in QL but I'm not sure about unintended side effects. And for some types, like date/time, we cannot do the conversion because a String might be a date math expression.

Reply:
Are there any new features supported by IN due to this new behavior? Such as date math?

Reply:
The only difference I know about is the missing conversion to the field type. This enables date math (and using string literal dates as in `date IN ('2021-10-01')`).

Reply:
> and `select emp_no from test_emp where emp_no < 9223372036854775807` fail

`<` is a BinaryComparison, but I guess for the `IN` rework I was expecting the `=` example :-)

Reply:
Good point, the `=` case also fails with a `query_shard_exception` :-)

Reply:
If the content of `IN` values is a `string` a query like `SELECT birth_date, birth_date IN ('now-68y/y', 'now-67y/y', '1958-09-05'::datetime), gender IN ('M'), birth_date IN ('1958-09-05'::datetime) FROM test_emp WHERE birth_date IN ('now-68y/y', 'now-67y/y', '1958-09-05') ORDER BY birth_date` will return `null` for `IN ('now-68y/y')`. The condition will be valid, thus assuming it returns `true`, but the actual value is `null` in the projection.

If the projection contains `birth_date IN (....'1958-09-05'::datetime)` the value will be `true`/`false` because of the  type that's the one expected by the field. But `WHERE birth_date IN ('1958-09-05')` (not a datetime explicitly) will still be `true`. That's inconsistent imo. Yes, we get the advantage of passing datemath for `IN` but the behavior is inconsistent between the condition and the projection.

Reply:
I agree that this is an inconsistency that should be addressed if possible but I'm not sure it should be done in this PR. The goal here is to make sure `a IN (x, y)` and `a = x OR a = y` behave consistently and the optimizer can freely turn one into the other.

The inconsistency between the condition and the projection is a necessary consequence because this is also how equality between datetime and string behaves:

```
SELECT birth_date = 'now-68y/y' from test_emp where birth_date = 'now-68y/y'

birth_date = 'now-68y/y'
------------------------
null
null
null
...
```

I would suggest to handle this in a separate issue though as this might be quite a rabbit hole.

Reply:
👍 see https://github.com/elastic/elasticsearch/issues/77055